src/nifreeze/data/base.py

```diff
@@ -0,0 +1,283 @@
+
+    def __getitem__(
+        self, idx: int | slice | tuple | np.ndarray
+    ) -> tuple[np.ndarray, np.ndarray | None]:
+        """
+        Returns volume(s) and corresponding affine(s) through fancy indexing.
+
+        Parameters
+        ----------
+        idx : :obj:`int` or :obj:`slice` or :obj:`tuple` or :obj:`~numpy.ndarray`
+            Indexer for the last dimension (or possibly other dimensions if extended).
+
+        Returns
+        -------
+        :obj:`tuple`
+            The selected volume(s) and corresponding affine(s).
+
+        """
+        if self.dataobj is None:
+            raise ValueError("No data available (dataobj is None).")
+
+        affine = self.motion_affines[idx] if self.motion_affines is not None else None
+        return self.dataobj[..., idx], affine
+
+    def get_filename(self) -> Path:
+        """Get the filepath of the HDF5 file."""
+        return self._filepath
+
+    def set_transform(self, index: int, affine: np.ndarray, order: int = 3) -> None:
+        """
+        Set an affine transform for a particular index and update the data object.
+
+        Parameters
+        ----------
+        index : :obj:`int`
+            The volume index to transform.
+        affine : :obj:`numpy.ndarray`
+            The 4x4 affine matrix to be applied.
+        order : :obj:`int`, optional
+            The order of the spline interpolation. Default is 3.
+
+        """
+        reference = namedtuple("ImageGrid", ("shape", "affine"))(
+            shape=self.dataobj.shape[:3], affine=self.affine
+        )
+
+        xform = Affine(matrix=affine, reference=reference)
+
+        if not Path(self._filepath).exists():
+            self.to_filename(self._filepath)
+
+        # read original DWI data & b-vector
+        with h5py.File(self._filepath, "r") as in_file:
+            root = in_file["/0"]
+            dataframe = np.asanyarray(root["dataobj"][..., index])
+
+        datamoving = nb.Nifti1Image(dataframe, self.affine, None)
+
+        # resample and update orientation at index
+        self.dataobj[..., index] = np.asanyarray(
+            xform.apply(datamoving, order=order).dataobj,
+            dtype=self.dataobj.dtype,
+        )
+
+        # if head motion affines are to be used, initialized to identities
+        if self.motion_affines is None:
+            self.motion_affines = np.repeat(np.eye(4)[None, ...], len(self), axis=0)
+
+        self.motion_affines[index] = xform.matrix
+
+    def to_filename(
+        self, filename: Path | str, compression: str | None = None, compression_opts: Any = None
+    ) -> None:
+        """
+        Write an HDF5 file to disk.
+
+        Parameters
+        ----------
+        filename : :obj:`os.pathlike`
+            The HDF5 file path to write to.
+        compression : :obj:`str`, optional
+            Compression filter ('gzip', etc.). Default is None (no compression).
+        compression_opts : :obj:`typing.Any`, optional
```

Comment:
Maybe the type used here is the right one, but just in case, seems like it must be an integer?
https://docs.h5py.org/en/stable/high/dataset.html#lossless-compression-filters